src/Telegram.Bot.Extensions.Polling/AsyncEnumerableReceivers/QueuedUpdateReceiver.cs

```diff
@@ -0,0 +1,182 @@
+﻿#if NETSTANDARD2_1
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Telegram.Bot.Requests;
+using Telegram.Bot.Types;
+using Telegram.Bot.Types.Enums;
+
+namespace Telegram.Bot.Extensions.Polling
+{
+    /// <summary>
+    /// Supports asynchronous iteration over <see cref="Update"/>s.
+    /// <para>Updates are received on a different thread and enqueued.</para>
+    /// </summary>
+    public class QueuedUpdateReceiver : IAsyncEnumerable<Update>
+    {
+        private readonly ITelegramBotClient _botClient;
+        private readonly UpdateType[]? _allowedUpdates;
+        private readonly Func<Exception, CancellationToken, Task>? _errorHandler;
+
+        private Enumerator? _enumerator;
+
+        /// <summary>
+        /// Constructs a new <see cref="QueuedUpdateReceiver"/> for the specified <see cref="ITelegramBotClient"/>
+        /// </summary>
+        /// <param name="botClient">The <see cref="ITelegramBotClient"/> used for making GetUpdates calls</param>
+        /// <param name="allowedUpdates">Indicates which <see cref="UpdateType"/>s are allowed to be received. null means all updates</param>
+        /// <param name="errorHandler">The function used to handle <see cref="Exception"/>s thrown by GetUpdates requests</param>
+        public QueuedUpdateReceiver(
+            ITelegramBotClient botClient,
+            UpdateType[]? allowedUpdates = default,
+            Func<Exception, CancellationToken, Task>? errorHandler = default)
+        {
+            _botClient = botClient ?? throw new ArgumentNullException(nameof(botClient));
+            _allowedUpdates = allowedUpdates;
+            _errorHandler = errorHandler;
+        }
+
+        /// <summary>
+        /// Indicates how many <see cref="Update"/>s are ready to be returned the enumerator
+        /// </summary>
+        public int PendingUpdates => _enumerator?.PendingUpdates ?? 0;
+
+        /// <summary>
+        /// Gets the <see cref="IAsyncEnumerator{Update}"/>. This method may only be called once.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> with which you can stop receiving</param>
+        public IAsyncEnumerator<Update> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+        {
+            var enumerator = new Enumerator(this, cancellationToken);
```

Comment:
If `GetAsyncEnumerator` will be called more than once this `Enumerator` instance will start to receive updates immediately interfering with the first enumerator. I think there should be additional `StartReceiving` method on the `Enumerator` that is called only when it's known that it's safe to start fetching updates.